nexus/db-model/src/instance.rs

```diff
@@ -220,6 +229,206 @@ impl InstanceRuntimeState {
+    /// configs permit them to reincarnate.
+    ///
+    /// Yes, this should probably be in `nexus-db-queries`, but it seemed nice
+    /// for it to be defined on the same struct as the in-memory logic
+    /// (`can_reincarnate`).
+    pub fn filter_reincarnatable(
+    ) -> impl diesel::query_builder::QueryFragment<pg::Pg>
+           + diesel::query_builder::QueryId
+           // All elements in this expression appear on the `instance` table, so
+           // it's a valid `filter` for that table, and the expression evaluates
+           // to a bool (or NULL), making it a valid WHERE clause.
+           + AppearsOnTable<instance::table, SqlType = Nullable<Bool>>
+           // I think this trait tells diesel that the query fragment has no
+           // GROUP BY clause, so that it knows it can be used as a WHERE clause
+           + ValidGrouping<(), IsAggregate = is_aggregate::No> {
+        use instance::dsl;
+
+        let now = diesel::dsl::now.into_sql::<pg::sql_types::Timestamptz>();
+
+        dsl::state
+            // Only attempt to restart Failed instances.
+            .eq(InstanceState::Failed)
+            // The instance's auto-restart policy must allow the control plane
+            // to restart it automatically.
+            //
+            // N.B. that this may become more complex in the future if we grow
+            // additional auto-restart policies that require additional logic
+            // (such as restart limits...)
+            .and(
+                dsl::auto_restart_policy
+                    .eq(InstanceAutoRestartPolicy::BestEffort),
+            )
+            // An instance whose last reincarnation was within the cooldown
+            // interval from now must remain in _bardo_ --- the liminal
+            // state between death and rebirth --- before its next
+            // reincarnation.
+            .and(
+                // If the instance has never previously been reincarnated, then
+                // it's allowed to reincarnate.
+                dsl::time_last_auto_restarted
+                    .is_null()
+                    // Or, if it has an overridden cooldown period, has that elapsed?
+                    .or(dsl::auto_restart_cooldown.is_not_null().and(
+                        dsl::time_last_auto_restarted
+                            .le(now.nullable() - dsl::auto_restart_cooldown),
+                    ))
+                    // Or, finally, if it does not have an overridden cooldown
+                    // period, has the default cooldown period elapsed?
+                    .or(dsl::auto_restart_cooldown.is_null().and(
+                        dsl::time_last_auto_restarted
+                            .le((now - Self::DEFAULT_COOLDOWN).nullable()),
+                    )),
+            )
+    }
+}
+
+/// It's just a type with the same representation as a `TimeDelta` that
+/// implements `Serialize` and `Deserialize`, because chrono apparently doesn't.
```

Comment:
It seems like chrono's `TimeDelta` does impl Serialize and Deserialize?
https://docs.rs/chrono/latest/chrono/struct.TimeDelta.html#impl-Serialize%3C__S%3E-for-TimeDelta

Reply:
It implements `Serialize`, but --- as far as I could tell from the documentation --- its implementation of `Deserialize` is not for `TimeDelta`, but [for the `rkyv::Archived` wrapper type](https://docs.rs/chrono/latest/chrono/struct.TimeDelta.html#impl-Deserialize%3CTimeDelta,+__D%3E-for-%3CTimeDelta+as+Archive%3E::Archived).

I went and added all of this stuff because just using a `TimeDelta` in a struct that derived `Deserialize` didn't compile. I think we might be able to remove our `Serialize` impl and just implement a custom `Deserialize`, but that felt a bit sketchy --- I was concerned that future semver-compatible changes to `chrono` could break our ability to roundtrip using their `Serialize` and our `Deserialize`. I can make the comment explain that more clearly.